Drop mypy excluded files

aciba90 opened the pull request:
## Proposed Commit Message
<!-- Include a proposed commit message because all PRs are squash merged -->

```
Drop mypy excluded files.

- Add types to let mypy pass.
- Add mypy flags:
  - detect unused ignores
  - redundant casts
- Drop support of `ConfigParser` in Python 2
- Harden DataSourceLXD.network_config
- Convert old-style commented types to proper types.
```

## Additional Context
<!-- If relevant -->

## Test Steps
<!-- Please include any steps necessary to verify (and reproduce if
this is a bug fix) this change on a live deployed system,
including any necessary configuration files, user-data,
setup, and teardown. Scripts used may be attached directly to this PR. -->
Run mypy in 3.6 <= Python <= 3.10

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes
that apply. -->
 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/contributing.html)
 - [x] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any documentation accordingly

aciba90 commented on cloudinit/sources/DataSourceLXD.py:
```diff
@@ -211,13 +203,19 @@ def network_config(self) -> dict:
         If none is present, then we generate fallback configuration.
         """
         if self._network_config == sources.UNSET:
```
Here, we are hardening the computation of `self.network_config` to work even if `self._get_data` wasn't called previously.

holmanb commented on cloudinit/serial.py:
```diff
@@ -1,12 +1,16 @@
+
 try:
-    from serial import Serial
+    from serial import Serial as _Serial
```
Looks like the try/except is from the 2.6 days. If this is still necessary, could we update the comments explaining why (in terms of python3)?

aciba90 replied:
Thanks for pointing out this @brettmilford. I have removed it and everything works.

holmanb replied:
Thanks!

holmanb reviewed (changes requested):
I just did a quick pass over this PR. The one inline comment about Serial import is that we might be able to drop the try/except and the FakeSerial class, and if not we should probably update the comment explaining why.

holmanb commented on cloudinit/config/cc_keyboard.py:
```diff
@@ -18,14 +18,14 @@
 DEFAULT_KEYBOARD_MODEL = "pc105"
 
-distros = distros.Distro.expand_osfamily(["arch", "debian", "redhat", "suse"])
+distros_ = distros.Distro.expand_osfamily(["arch", "debian", "redhat", "suse"])
```
Nitpick: Why the underscore here? Is this to disambiguate from the import? The trailing underscore is a bit unusual, maybe we could use `supported_distros` or something like that instead?

aciba90 replied:
Exactly, the intention was to differentiate this variable name from the import. mypy doesn't like the same variable being a module and a list. It is also better like this if a future modification requires using the `distro` module.

Trailing underscore is used by convention to avoid conflicts with Python keyword: [PEP8](https://peps.python.org/pep-0008/#descriptive-naming-styles), which is not the case here.

Variable name improved. Thanks for this suggestion!

holmanb reviewed (approved):
Thanks for this! I don't see anything requiring change, but I'll leave this unmerged at the moment in case @blackboxsw or @TheRealFalcon have anything to add.

holmanb commented:
Close/reopen bump to rerun integration tests. It appears snap install failed.

TheRealFalcon reviewed (changes requested):
Thanks @aciba90 . It'll be great to have the mypy checks everywhere. I left some comments inline. Not everything is super major, so feel free to push back.

Also, the `# type: ignore` annotations are added to silence pyright not mypy. Pyright isn't required or part of our build, but if mypy doesn't complain about it, I think I'd rather keep them because pyright is much more strict in its checking.

aciba90 commented on tests/unittests/cmd/test_cloud_id.py:
```diff
@@ -15,7 +15,7 @@
 class TestCloudId:
 
-    args = namedtuple("cloudidargs", "instance_data json long")
+    CloudIdArgs = namedtuple("CloudIdArgs", "instance_data json long")
```
Removed

aciba90 commented on tests/unittests/cmd/test_main.py:
```diff
@@ -13,8 +13,8 @@
 
-mypaths = namedtuple("MyPaths", "run_dir")
-myargs = namedtuple("MyArgs", "debug files force local reporter subcommand")
+MyPaths = namedtuple("MyPaths", "run_dir")
```
Removed

aciba90 commented on tests/unittests/sources/test_lxd.py:
```diff
@@ -51,12 +51,27 @@ def _add_network_v1_device(devname) -> dict:
     return LXD_V1_METADATA
 
 
+@pytest.fixture
```
Removed

aciba90 commented on cloudinit/cmd/query.py:
```diff
@@ -150,7 +150,6 @@ def _read_instance_data(instance_data, user_data, vendor_data) -> dict:
     :raise: IOError/OSError on absence of instance-data.json file or invalid
         access perms.
     """
-    paths = None
```
In theory yes, but having a deeper look into the logic of the function we can see that the `paths` var is only accessed in the else branches of `if instance_data:`, `if user_data:` and `if vendor_data:`. That is why we only compute paths `if not all([instance_data, user_data, vendor_data]):`. Therefore paths is never going to be undefined. Furthermore, the function assumes `paths` is an instance of `Paths` which would also fail if we keep `paths = None`.

Mypy is not clever enough to realize this, which could be a symptom of complexity in this function. I though the more pragmatic approach for this PR is to just remove the unneeded var definition. But please let me know if you think this need more work.

aciba90 commented on cloudinit/handlers/jinja_template.py:
```diff
@@ -4,15 +4,19 @@
 import re
 from errno import EACCES
-from typing import Optional
+from typing import Optional, Type
```
I agree this version is more compact and legible, but we need an extra line to avoid flake8 think we are doing logic before the subsequent imports and mark them as `E402 module level import not at top of file`. That's why we need to add the `TYPE_CHECKING` check:

```python
if TYPE_CHECKING:
    JUndefinedError: Type[Exception]
try:
    from jinja2.exceptions import UndefinedError as JUndefinedError
    from jinja2.lexer import operator_re
except ImportError:
    # No jinja2 dependency
    JUndefinedError = Exception
    operator_re = re.compile(r"[-.]")
```

I am okay with both syntax.

TheRealFalcon replied:
Yours is ok, or we could move the try to the bottom of the imports. isort and black shouldn't complain because they're in a try/except. Either way is fine. You can choose.

aciba90 replied:
Your version moved down. Thanks!

aciba90 commented on cloudinit/templater.py:
```diff
@@ -13,15 +13,19 @@
     from jinja2 import Template as JTemplate
 
-    JINJA_AVAILABLE = True
+    JUndefined: Type
```
Same answer.

aciba90 commented on cloudinit/sources/DataSourceLXD.py:
```diff
@@ -211,13 +203,19 @@ def network_config(self) -> dict:
                     "network-config"
                 )
-            else:
+                if isinstance(crawled_network_config, dict):
```
If we can ensure that the content of `self._crawled_metadata["network-config"]` is always a dict, then this check could be promoted to a cast (to help mypy) and remove the runtime overhead (I am assuming performance is your concern).

But if we remove the check and assign something different than a dict or `sources.UNSET` to `network_config`, then we are going to skip the generation of the fallback configuration, and probably fail in client code.

Maybe I am being over defensive in this code. Please let me know your thoughts, thanks.

TheRealFalcon replied:
My bigger concern is readability and `isinstance()`  checks everywhere where we could instead just check for UNSET, but it is fine as-is.

I think these `UNSET` variables aren't really needed and we could instead just use `None` to simplify a lot of the typing, but that's outside the context of this PR.

aciba90 replied:
Simplified using None. I wasn't sure that this was possible. Thanks for the suggestion.

aciba90 commented:
Thanks @TheRealFalcon for the review and pointing out improvements. I have addressed most of them and left you a couple of comments.

> Also, the `# type: ignore` annotations are added to silence pyright not mypy. Pyright isn't required or part of our build, but if mypy doesn't complain about it, I think I'd rather keep them because pyright is much more strict in its checking.

I have changed them to `# pyright: ignore` to make clear that they are `pyright` specific and let `mypy` warn about unused `# type: ignore`s. More info: [Pyright doc](https://github.com/microsoft/pyright/blob/main/docs/comments.md#line-level-diagnostic-suppression).

This PR is ready to review.

TheRealFalcon reviewed (approved):
LGTM, will merge assuming CI passes.

aciba90 commented:
Try/excepts moved down and `UNSET` removed from `DataSourceLXD`. Thanks @TheRealFalcon !
